pkg/ccl/changefeedccl/kvevent/blocking_buffer.go

```diff
@@ -38,35 +38,67 @@ type blockingBuffer struct {
 
 	mu struct {
 		syncutil.Mutex
-		closed  bool                   // True when buffer closed.
-		reason  error                  // Reason buffer is closed.
-		drainCh chan struct{}          // Set when Drain request issued.
-		blocked bool                   // Set when event is blocked, waiting to acquire quota.
-		queue   *bufferEventChunkQueue // Queue of added events.
+		closed     bool                   // True when buffer closed.
+		reason     error                  // Reason buffer is closed.
+		drainCh    chan struct{}          // Set when Drain request issued.
+		numBlocked int                    // Number of waitors blocked to acquire quota.
+		queue      *bufferEventChunkQueue // Queue of added events.
 	}
 }
 
 // NewMemBuffer returns a new in-memory buffer which will store events.
 // It will grow the bound account to buffer more messages but will block if it
 // runs out of space. If ever any entry exceeds the allocatable size of the
 // account, an error will be returned when attempting to buffer it.
-func NewMemBuffer(
-	acc mon.BoundAccount, sv *settings.Values, metrics *Metrics, opts ...quotapool.Option,
+func NewMemBuffer(acc mon.BoundAccount, sv *settings.Values, metrics *Metrics) Buffer {
+	return newMemBuffer(acc, sv, metrics, nil)
+}
+
+// TestingNewMemBuffer allows test to construct buffer which will invoked
+// specified notification function when blocked, waiting for memory.
+func TestingNewMemBuffer(
+	acc mon.BoundAccount,
+	sv *settings.Values,
+	metrics *Metrics,
+	onWaitStart quotapool.OnWaitStartFunc,
 ) Buffer {
-	const slowAcquisitionThreshold = 5 * time.Second
+	return newMemBuffer(acc, sv, metrics, onWaitStart)
+}
 
-	opts = append(opts,
-		quotapool.OnSlowAcquisition(slowAcquisitionThreshold, logSlowAcquisition(slowAcquisitionThreshold)),
-		quotapool.OnWaitFinish(
-			func(ctx context.Context, poolName string, r quotapool.Request, start time.Time) {
-				metrics.BufferPushbackNanos.Inc(timeutil.Since(start).Nanoseconds())
-			}))
+func newMemBuffer(
+	acc mon.BoundAccount,
+	sv *settings.Values,
+	metrics *Metrics,
+	onWaitStart quotapool.OnWaitStartFunc,
+) Buffer {
+	const slowAcquisitionThreshold = 5 * time.Second
 
 	b := &blockingBuffer{
 		signalCh: make(chan struct{}, 1),
 		metrics:  metrics,
 		sv:       sv,
 	}
+
+	opts := []quotapool.Option{
+		quotapool.OnSlowAcquisition(slowAcquisitionThreshold, logSlowAcquisition(slowAcquisitionThreshold)),
+		quotapool.OnWaitStart(func(ctx context.Context, poolName string, r quotapool.Request) {
+			if onWaitStart != nil {
+				onWaitStart(ctx, poolName, r)
+			}
+			b.mu.Lock()
+			b.mu.numBlocked++
+			b.mu.Unlock()
+			b.notifyOutOfQuota()
```

Comment:
The question I have is whether this can lead to back-to-back emission of flush events. I suppose that's fine, especially now, but in a world with some processing parallelism it might get weird. 

Can we come up with something simple to prevent that? Either bookkeeping or more careful synchronization?

Reply:
> The question I have is whether this can lead to back-to-back emission of flush events. I suppose that's fine, especially now, but in a world with some processing parallelism it might get weird.
> 
> Can we come up with something simple to prevent that? Either bookkeeping or more careful synchronization?

I don't think this can lead to back-to-back emission of flush events, with exception of one corner case.
I think there are few important observations:
  1. signalCh has a buffer of 1; this buffer is to ensure that notification to single-threaded consumer (Get) never misses notification.
  2. Producers (there can be multiple, e.g. during backfill) may concurrently run out of quota; however only 1 of those produces will be at the head of the (quota pool) queue, and thus only 1 consumer will invoke notify out of quota
  3. (This really flows from 1 & 2) -- Producer either puts an event into blocking buffer queue, *OR* it puts a notification onto the channel that it is out of quota.

With this in mind, the condition to trigger Flush is: the queue is empty *and* we have some producers blocked
(`if !ok && b.mu.numBlocked > 0 ...`).  That means that producer is blocked and there is nothing else in blocking buffer queue to cause consumer to try to release any resources -- i.e. every allocated resource is buffered.
So, we trigger flush.  Because this flush was triggered when there were no outstanding events in blocking buffer
queue, that means that *every*  outstanding event/allocations must have been released.
The original producer that was blocked will attempt to acquire resources, now that everything has been released.
If the original producer failed to acquire resource (that means that our parent memory monitor is *very* tight), then that's fine -- we will trigger a no-op flush -- and we'll do that once a second (that's the corner case).

If it acquired resource -- great, it queues the event.  We have made forward progress; if the next producer also blocked -- well, that means that perhaps our memory buffer is tiny, or events are huge -- either way flush will be triggered correctly -- even if that means that we flush 1 event.

Is my analysis sound? Do you still think more book-keeping needed?  Or perhaps just copying above as a comment because it's not obvious at all.

Reply:
1) Publisher enqueues and is blocked, so sends on the channel
2) Consumer is signaled by the send, locks the mutex, notices that there's a blocked producer, creates a flush event
3) Second publisher enqueues and sends on the channel (or not)
4) Consumer now sends another flush event

Reply:
> * Publisher enqueues and is blocked, so sends on the channel
> * Consumer is signaled by the send, locks the mutex, notices that there's a blocked producer, creates a flush event
> * Second publisher enqueues and sends on the channel (or not)
> * Consumer now sends another flush event

I don't think 3 can happen; because all blocked producers are blocked in a queue of their own (in quota pool); thus; 
after step 2 (Flush) the only producer that can wake up is the first producer -- and that producer either produces
an event or is blocked again (as described in the corner case above).

Reply:
I stand corrected; I guess a no-op flush is possible.